Source/VCGeneration/Counterexample.cs

```diff
@@ -386,6 +391,98 @@ public Model.Element GetModelValue(Variable v)
     }
 
     public abstract int GetLocation();
+
+    public ErrorInformation CreateErrorInformation(ConditionGeneration.Outcome outcome, bool forceBplErrors)
+    {
+      ErrorInformation errorInfo;
+      var cause = "Error";
+      if (outcome == VCGen.Outcome.TimedOut)
+      {
+        cause = "Timed out on";
+      }
+      else if (outcome == VCGen.Outcome.OutOfMemory)
+      {
+        cause = "Out of memory on";
+      }
+      else if (outcome == VCGen.Outcome.SolverException)
+      {
+        cause = "Solver exception on";
+      }
+      else if (outcome == VCGen.Outcome.OutOfResource)
+      {
+        cause = "Out of resource on";
+      }
+
+      if (this is CallCounterexample callError)
+      {
+        if (callError.FailingRequires.ErrorMessage == null || forceBplErrors)
+        {
+          errorInfo = ErrorInformationFactory.Instance.CreateErrorInformation(callError.FailingCall.tok,
+            callError.FailingCall.ErrorData as string ?? callError.FailingCall.Description.FailureDescription,
+            cause);
+          errorInfo.Kind = ErrorKind.Precondition;
+          errorInfo.AddAuxInfo(callError.FailingRequires.tok,
+            callError.FailingRequires.ErrorData as string ?? callError.FailingRequires.Description.FailureDescription,
+            "Related location");
+        }
+        else
+        {
+          errorInfo = ErrorInformationFactory.Instance.CreateErrorInformation(null,
+            callError.FailingRequires.ErrorMessage);
+        }
+      }
+      else if (this is ReturnCounterexample returnError)
+      {
+        if (returnError.FailingEnsures.ErrorMessage == null || forceBplErrors)
+        {
+          errorInfo = ErrorInformationFactory.Instance.CreateErrorInformation(returnError.FailingReturn.tok,
+            returnError.FailingReturn.Description.FailureDescription);
+          errorInfo.Kind = ErrorKind.Postcondition;
+          errorInfo.AddAuxInfo(returnError.FailingEnsures.tok,
+            returnError.FailingEnsures.ErrorData as string ?? returnError.FailingEnsures.Description.FailureDescription,
+            "Related location");
+        }
+        else
+        {
+          errorInfo = ErrorInformationFactory.Instance.CreateErrorInformation(null,
+            returnError.FailingEnsures.ErrorMessage);
+        }
+      }
+      else // error is AssertCounterexample
+      {
+        Debug.Assert(this is AssertCounterexample);
+        var assertError = (AssertCounterexample)this;
```

Comment:
This line will already throw an invalid-cast exception if the type isn't right; do you need the Debug.Assert above?

Reply:
Removed.